### ISSUE TITLE
M1 macOS build?

### ISSUE THREAD

**Issue by Willian-Zhang** [transmission-remote-gui/transgui#1355]
Apple m1 lines Macs are out, is there a easy way to build a m1 native app?

Operating system: macOS Big Sur 11.0.1 (20B29)

**Comment by Willian-Zhang** [comment]
Since there is no building guide provided for macOS,  https://github.com/transmission-remote-gui/transgui/pull/1224/files and https://wiki.freepascal.org/macOS_Big_Sur_changes_for_developers#ARM64.2FAArch64.2FAppleSilicon_Support might be helpful for reference.

**Comment by PeterDaveHello** [comment]
@Willian-Zhang I don't think we have resource to work on it yet, we'll need some help to do that.

**Comment by thebitstick** [comment]
XCode 12 has cross-compiling, so you can build for ARM on an x86 Mac.
I'd be happy to test builds with my M1 Mac Mini.

**Comment by PeterDaveHello** [comment]
@thebitstick Thanks! Contributions are very welcome and needed here!

**Comment by flying-sausages** [comment]
And my axe! If there are clear enough setup guides I'd be happy to offer my resources, both machine and time wise

**Comment by shermanikk** [comment]
I've managed to rebuild a couple apps like ffmpeg, mpv, and iina for arm64 before any official builds were released so I'd be more than willing to give this a go and do a little how-to writeup. This and Dropbox are the only x86 apps left that I use on a daily basis.

**Comment by kellertuer** [comment]
I am completely inexperienced in such things, but if you need another tester and have a dmg or something I can surely also test it on my M1.

**Comment by shermanikk** [comment]
Yeah I've got an M1 MBA myself, hence the motivation to do these pre-release arm64 builds. I won't be able to really put my head into it until later today but anything I find out will be posted here.

> I am completely inexperienced in such things, but if you need another tester and have a dmg or something I can surely also test it on my M1.

**Comment by zepretender** [comment]
I don't know how to compile a project... But I have a Mac Mini M1 now and I really miss Transmission Remote GUI :) 
If I can help, I'll do !

**Comment by kellertuer** [comment]
Well the old Transmission Remote GUI works fine on my M1 with Rosetta; just since it only respects dark mode half, it does not look nice.

**Comment by zepretender** [comment]
> Well the old Transmission Remote GUI works fine on my M1 with Rosetta; just since it only respects dark mode half, it does not look nice.

How do you do ? Here, I can't launch it. It's "PowerPC/32bits", so it can't be launched

**Comment by kellertuer** [comment]
Are you using 5.18? That is 64 bit I think? Also Homebrew _might_ work https://formulae.brew.sh/cask/transmission-remote-gui

**Comment by zepretender** [comment]
> Are you using 5.18? That is 64 bit I think? Also Homebrew _might_ work https://formulae.brew.sh/cask/transmission-remote-gui

Yes, I've downloaded latest release (5.18). It doesn't launch. When I use iMobie M1 App Checker, it says that this release can't be launch on a M1.

For the homebrew... I'm to newbie to understand how to do :(

**Comment by valentine** [comment]
@zepretender You might be encountering issues with the app [not being signed](https://support.apple.com/en-us/HT202491). Go to your Applications folder, right-click on the app, and select Open. You'll only have to do this once.

**Comment by shermanikk** [comment]
<img width="377" alt="Screen Shot 2021-01-19 at 6 01 46 AM" src="https://user-images.githubusercontent.com/2495495/105044684-dd660680-5a1b-11eb-9357-efbb6ca13aa5.png">

Okay, after a bunch of fiddling I have managed to successfully build transgui for aarch64. Hooray! Unfortunately, it crashes on launch with this error:

`Process:               transgui [71407]
Path:                  /Volumes/*/Transmission Remote GUI.app/Contents/MacOS/transgui
Identifier:            com.transgui
Version:               Transmission Remote GUI 5.18.0 (5.18.0)
Code Type:             ARM-64 (Native)
Parent Process:        ??? [1]
Responsible:           transgui [71407]
User ID:               501

Date/Time:             2021-01-19 06:02:50.382 -0800
OS Version:            macOS 11.1 (20C69)
Report Version:        12
Anonymous UUID:        5ADF9579-9D2E-A80E-3DD2-31335748639B

Sleep/Wake UUID:       5E8D81A4-FF06-4106-960A-511FE02286A4

Time Awake Since Boot: 50000 seconds
Time Since Wake:       1300 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
/usr/lib/libcrypto.dylib
abort() called
Invalid dylib load. Clients should not load the unversioned libcrypto dylib as it does not have a stable ABI.`

Now I'm not that familiar with the transgui source, or pascal at all really, so while I could probably eventually figure out how to get the linker to use the openssl 1.1.1 dylibs installed by homebrew instead of the deprecated ones it's trying to use, I figured I would save myself a whole ton of time and frustration by at least asking if someone here with more transgui/pascal experience could point me in the right direction. I'm assuming I have to to two things: add the /opt/homebrew/Cellar/openssl@1.1/1.1.1i/ directory to the ld search path and edit ssl_openssl_lib.pas lines 116 and 117 to be

`    DLLSSLName: string = 'libssl.1.1.dylib';
    DLLUtilName: string = 'libcrypto.1.1.dylib';`

instead of

`    DLLSSLName: string = 'libssl.dylib';
    DLLUtilName: string = 'libcrypto.dylib';`

also probably want to static link them so that end-users don't have to have homebrew openssl installed on their machines to run transgui, right?

There will probably be more issues after this, but that's where I'm at right now.

**Comment by shermanikk** [comment]
Oh and I almost forgot, for anyone else who wants to join me in the mudpit, here's what I've had to do to get to this point:

>1) Build FPC from latest git source
>2) Build Lazarus from trunk
>3) Install openssl with homebrew
>4) Add to .fpc.cfg:
		`-Fl/opt/homebrew/Cellar/openssl@1.1/1.1.1i/lib`
>5) Edit transgui/synapse/source/lib/ssl_openssl_lib.pas:
>
```
@@ -113,8 +113,8 @@ const
		 var
		   {$IFNDEF MSWINDOWS}
		     {$IFDEF DARWIN}
		-    DLLSSLName: string = 'libssl.dylib';
		-    DLLUtilName: string = 'libcrypto.dylib';
		+    DLLSSLName: string = 'libssl.1.1.dylib';^M
		+    DLLUtilName: string = 'libcrypto.1.1.dylib';^M
		     {$ELSE}
		      {$IFDEF OS2}
		       {$IFDEF OS2GCC}
```
>6) Edit setup/macosx/create_app_new.sh:
```
                -lazdir="${1:-/Developer/lazarus/}"
		+lazdir="${1:-<your laz_trunk dir>}

		 if [ -z "${CI-}" ]; then
		   ./install_deps.sh
		@@ -24,11 +24,11 @@ fi
		 mkdir -p ../../Release/
		 sed -i.bak "s/'Version %s'/'Version %s Build $build'#13#10'Compiled by: $fpc_ver, Lazarus v$lazarus_ver'/" ../../about.lfm
		 
		-lazbuild -B ../../transgui.lpi --lazarusdir=/Developer/lazarus/ --compiler=/usr/local/bin/fpc --cpu=x86_64 --widgetset=cocoa
		+lazbuild -B ../../transgui.lpi --lazarusdir=<your laz_trunk dir> --compiler=/usr/local/bin/fpc --cpu=aarch64 --widgetset=cocoa
		 
		 # Building Intel version
		-make -j"$(sysctl -n hw.ncpu)" -C ../.. clean CPU_TARGET=x86_64 "$lazdir"
		-make -j"$(sysctl -n hw.ncpu)" -C ../.. CPU_TARGET=x86_64 "$lazdir"
		+make -j"$(sysctl -n hw.ncpu)" -C ../.. clean CPU_TARGET=aarch64 "$lazdir"
		+make -j"$(sysctl -n hw.ncpu)" -C ../.. CPU_TARGET=aarch64 "$lazdir" 
```
>7) Edit restranslator.pas

```@@ -784,8 +784,9 @@ var
		   ResourceName: AnsiString;
		   OwnerName: AnsiString;
		 begin
		-  if Sender is TReader and Assigned(TReader(Sender).Owner) then
		-    OwnerName := TReader(Sender).Owner.GetNamePath;
		+  if Sender is TReader {and Assigned(TReader(Sender).Owner)} then
		+    if Assigned(TReader(Sender).Owner) then
		+      OwnerName := TReader(Sender).Owner.GetNamePath;
```
>8) cd setup/macosx
>9) bash create_app_new.sh
>10) **_\*jeff goldblum voice\* Step 10? There is no step 10!_**

**Comment by kellertuer** [comment]
Thanks, this already looks quite great! (looking forward to finally being able to read marked lines in Dark mode).
On my machine, sadly, your provided app can not be launched (the application is damaged and not be started is the error message)

**Comment by shermanikk** [comment]
Oh right, it's not really signed since it's not XCode. 

> Thanks, this already looks quite great! (looking forward to finally being able to read marked lines in Dark mode).
> On my machine, sadly, your provided app can not be launched (the application is damaged and not be started is the error message)

Huh, I definitely signed it with my dev account so anyone should be able to run it. What happens when you try to run "Transmission Remote GUI.app/Contents/MacOS/transgu" directly?

**Comment by kellertuer** [comment]
No, sorry, I know what an unsigned application looks like and how to start it. I mean the app is broken like in: I can not open the container to access the binary you refer to.

**Comment by shermanikk** [comment]
> No, sorry, I know what an unsigned application looks like and how to start it. I mean the app is broken like in: I can not open the container to access the binary you refer to.

That's bizarre, it definitely had the proper structure when I zipped it, and I tried downloading the zip on another machine and it unzipped fine. Either way I signed the entire .app bundle and re-uploaded the zip, so try downloading it again.

**Comment by kellertuer** [comment]
I still got a “signed by untrusted developer”, but that's easy and is only the first start, so that works! Nice! Finally a proper dark mode and a native app. Thanks!

**Comment by shermanikk** [comment]
> I still got a “signed by untrusted developer”, but that's easy and is only the first start, so that works! Nice! Finally a proper dark mode and a native app. Thanks!

Awesome, I guess that zip got messed up somehow when I uploaded it or something, weird. Glad it's working for others now, if I did all that work and nobody else could benefit it'd break my poor little nerd heart. Hooray!

**Comment by thebitstick** [comment]
Tested @shermanikk's build and it runs great.
Great job!

**Comment by jeromeof** [comment]
Great work @shermanikk - but I oddly get an error "Unable to load OpenSSL library files: libssl.1.1.dylib and libcrypto.1.1.dylib" after giving the app permissions. It sort of runs still but cannot connect to my Transmission Daemon running over SSL

**Comment by shermanikk** [comment]
> 
> 
> Great work @shermanikk - but I oddly get an error "Unable to load OpenSSL library files: libssl.1.1.dylib and libcrypto.1.1.dylib" after giving the app permissions. It sort of runs still but cannot connect to my Transmission Daemon running over SSL

Oh I guess they didn't get built-in, still new at this. All you need to do is install openssl with homebrew in /opt/homebrew. That's the location I used when building it so I'm assuming it will be looking for them there.

**Comment by qinyongqiang** [comment]
> > Great work @shermanikk - but I oddly get an error "Unable to load OpenSSL library files: libssl.1.1.dylib and libcrypto.1.1.dylib" after giving the app permissions. It sort of runs still but cannot connect to my Transmission Daemon running over SSL
> 
> Oh I guess they didn't get built-in, still new at this. All you need to do is install openssl with homebrew in /opt/homebrew. That's the location I used when building it so I'm assuming it will be looking for them there.

Thank you for your fantastic work! I also encounter the openssl issue too... I tried to compile the code from scratch but now is blocked while linking transgui... The ld complains about 
```
(9015) Linking ./transgui
ld: warning: building for macOS 10.5 is deprecated
ld: warning: passed two min versions (10.5, 11.0.0) for platform macOS. Using 11.0.0.
ld: library not found for -lc
An error occurred while linking
Error: (9013) Error while linking
Fatal: (10026) There were 1 errors compiling module, stopping
Fatal: (1018) Compilation aborted
```
Any ideas? I used the official fpc 3.2.2 for M1 and compiled Lazarus myself. Other configurations are following your instructions.

**Comment by flying-sausages** [comment]
> Oh I guess they didn't get built-in, still new at this. All you need to do is install openssl with homebrew in /opt/homebrew. That's the location I used when building it so I'm assuming it will be looking for them there.

Which version were you using? Mine cannot be found. This is my install of OpenSSL via brew.

```
~ ❯❯❯ brew info openssl
openssl@1.1: stable 1.1.1k (bottled) [keg-only]
Cryptography and SSL/TLS Toolkit
https://openssl.org/
/opt/homebrew/Cellar/openssl@1.1/1.1.1k (8,071 files, 18MB)
  Poured from bottle on 2021-03-29 at 19:37:47
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openssl@1.1.rb
License: OpenSSL
==> Caveats
A CA file has been bootstrapped using certificates from the system
keychain. To add additional certificates, place .pem files in
  /opt/homebrew/etc/openssl@1.1/certs

and run
  /opt/homebrew/opt/openssl@1.1/bin/c_rehash

openssl@1.1 is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS provides LibreSSL.

If you need to have openssl@1.1 first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc

For compilers to find openssl@1.1 you may need to set:
  export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
  export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"

For pkg-config to find openssl@1.1 you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"

==> Analytics
install: 748,044 (30 days), 2,616,398 (90 days), 8,792,644 (365 days)
install-on-request: 55,366 (30 days), 240,031 (90 days), 1,126,038 (365 days)
build-error: 0 (30 days)
```

**Comment by qinyongqiang** [comment]
> > Oh I guess they didn't get built-in, still new at this. All you need to do is install openssl with homebrew in /opt/homebrew. That's the location I used when building it so I'm assuming it will be looking for them there.
> 
> Which version were you using? Mine cannot be found. This is my install of OpenSSL via brew.
> 
> ```
> ~ ❯❯❯ brew info openssl
> openssl@1.1: stable 1.1.1k (bottled) [keg-only]
> Cryptography and SSL/TLS Toolkit
> https://openssl.org/
> /opt/homebrew/Cellar/openssl@1.1/1.1.1k (8,071 files, 18MB)
>   Poured from bottle on 2021-03-29 at 19:37:47
> From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openssl@1.1.rb
> License: OpenSSL
> ==> Caveats
> A CA file has been bootstrapped using certificates from the system
> keychain. To add additional certificates, place .pem files in
>   /opt/homebrew/etc/openssl@1.1/certs
> 
> and run
>   /opt/homebrew/opt/openssl@1.1/bin/c_rehash
> 
> openssl@1.1 is keg-only, which means it was not symlinked into /opt/homebrew,
> because macOS provides LibreSSL.
> 
> If you need to have openssl@1.1 first in your PATH, run:
>   echo 'export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
> 
> For compilers to find openssl@1.1 you may need to set:
>   export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
>   export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
> 
> For pkg-config to find openssl@1.1 you may need to set:
>   export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"
> 
> ==> Analytics
> install: 748,044 (30 days), 2,616,398 (90 days), 8,792,644 (365 days)
> install-on-request: 55,366 (30 days), 240,031 (90 days), 1,126,038 (365 days)
> build-error: 0 (30 days)
> ```

Seems like same to yours. (except the brew source)
```
$ brew info openssl
openssl@1.1: stable 1.1.1k (bottled) [keg-only]
Cryptography and SSL/TLS Toolkit
https://openssl.org/
/opt/homebrew/Cellar/openssl@1.1/1.1.1k (8,071 files, 18MB)
  Poured from bottle on 2021-05-21 at 14:34:05
From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/openssl@1.1.rb
License: OpenSSL
==> Caveats
A CA file has been bootstrapped using certificates from the system
keychain. To add additional certificates, place .pem files in
  /opt/homebrew/etc/openssl@1.1/certs

and run
  /opt/homebrew/opt/openssl@1.1/bin/c_rehash

openssl@1.1 is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS provides LibreSSL.

If you need to have openssl@1.1 first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc

For compilers to find openssl@1.1 you may need to set:
  export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
  export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"

For pkg-config to find openssl@1.1 you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"

==> Analytics
install: 748,044 (30 days), 2,616,398 (90 days), 8,792,644 (365 days)
install-on-request: 55,366 (30 days), 240,031 (90 days), 1,126,038 (365 days)
build-error: 0 (30 days)

```

**Comment by shermanikk** [comment]
> > > Oh I guess they didn't get built-in, still new at this. All you need to do is install openssl with homebrew in /opt/homebrew. That's the location I used when building it so I'm assuming it will be looking for them there.
> > 
> > 
> > Which version were you using? Mine cannot be found. This is my install of OpenSSL via brew.
> > ```
> > ~ ❯❯❯ brew info openssl
> > openssl@1.1: stable 1.1.1k (bottled) [keg-only]
> > Cryptography and SSL/TLS Toolkit
> > https://openssl.org/
> > /opt/homebrew/Cellar/openssl@1.1/1.1.1k (8,071 files, 18MB)
> >   Poured from bottle on 2021-03-29 at 19:37:47
> > From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/openssl@1.1.rb
> > License: OpenSSL
> > ==> Caveats
> > A CA file has been bootstrapped using certificates from the system
> > keychain. To add additional certificates, place .pem files in
> >   /opt/homebrew/etc/openssl@1.1/certs
> > 
> > and run
> >   /opt/homebrew/opt/openssl@1.1/bin/c_rehash
> > 
> > openssl@1.1 is keg-only, which means it was not symlinked into /opt/homebrew,
> > because macOS provides LibreSSL.
> > 
> > If you need to have openssl@1.1 first in your PATH, run:
> >   echo 'export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
> > 
> > For compilers to find openssl@1.1 you may need to set:
> >   export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
> >   export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
> > 
> > For pkg-config to find openssl@1.1 you may need to set:
> >   export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"
> > 
> > ==> Analytics
> > install: 748,044 (30 days), 2,616,398 (90 days), 8,792,644 (365 days)
> > install-on-request: 55,366 (30 days), 240,031 (90 days), 1,126,038 (365 days)
> > build-error: 0 (30 days)
> > ```
> 
> Seems like same to yours. (except the brew source)
> 
> ```
> $ brew info openssl
> openssl@1.1: stable 1.1.1k (bottled) [keg-only]
> Cryptography and SSL/TLS Toolkit
> https://openssl.org/
> /opt/homebrew/Cellar/openssl@1.1/1.1.1k (8,071 files, 18MB)
>   Poured from bottle on 2021-05-21 at 14:34:05
> From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/openssl@1.1.rb
> License: OpenSSL
> ==> Caveats
> A CA file has been bootstrapped using certificates from the system
> keychain. To add additional certificates, place .pem files in
>   /opt/homebrew/etc/openssl@1.1/certs
> 
> and run
>   /opt/homebrew/opt/openssl@1.1/bin/c_rehash
> 
> openssl@1.1 is keg-only, which means it was not symlinked into /opt/homebrew,
> because macOS provides LibreSSL.
> 
> If you need to have openssl@1.1 first in your PATH, run:
>   echo 'export PATH="/opt/homebrew/opt/openssl@1.1/bin:$PATH"' >> ~/.zshrc
> 
> For compilers to find openssl@1.1 you may need to set:
>   export LDFLAGS="-L/opt/homebrew/opt/openssl@1.1/lib"
>   export CPPFLAGS="-I/opt/homebrew/opt/openssl@1.1/include"
> 
> For pkg-config to find openssl@1.1 you may need to set:
>   export PKG_CONFIG_PATH="/opt/homebrew/opt/openssl@1.1/lib/pkgconfig"
> 
> ==> Analytics
> install: 748,044 (30 days), 2,616,398 (90 days), 8,792,644 (365 days)
> install-on-request: 55,366 (30 days), 240,031 (90 days), 1,126,038 (365 days)
> build-error: 0 (30 days)
> ```

I'm totally new to pascal as well so if anyone can help me with this too, I'd be super grateful. You'll notice in my original comment I did something stupid: I pointed directly to version 1.1.1i of the openssl libs in my .fpc.cfg file. So when homebrew updated openssl, this obviously broke. What I don't understand is I have tried every fucking way possible of recompiling starting from SCRATCH with not even fpc installed and no matter what every version of transgui I manage to build always looks for openssl in that directory. It's like my .fpc.cfg file means nothing anymore.

As a temporary workaround making a symlink of the current version to 1.1.1i in /opt/homebrew/Cellar/openssl@1.1/ does work for me, so try that. It's not fixing the core issue though which I am totally stumped by.

**Comment by shermanikk** [comment]
Okay, figured out something at least. I can include the dylibs with the .app directory so you don't even have to have homebrew install OpenSSL! All I needed to do was copy libcrypto.1.1.dylib and libssl.1.1.dylib into the Contents/MacOS/ directory inside the .app bundle, then run "install_name_tool -add_rpath @executable_path/. transgui" inside the same directory. I've uploaded this self-contained version, so if someone could try it on their own machine without OpenSSL installed to make sure it works for them that'd be great: https://shermanikk.net/stuff/transgui-macosx-arm64.zip

**Comment by flying-sausages** [comment]
>  https://shermanikk.net/stuff/transgui-macosx-arm64.zip

I tried opening through the right-click menu but it wouldn't budge.

<img width="372" alt="image" src="https://user-images.githubusercontent.com/23618693/120921426-ec480f80-c6c3-11eb-8b91-75b5433d8a33.png">

**Comment by shermanikk** [comment]
> > https://shermanikk.net/stuff/transgui-macosx-arm64.zip
> 
> I tried opening through the right-click menu but it wouldn't budge.
> 
> <img alt="image" width="372" src="https://user-images.githubusercontent.com/23618693/120921426-ec480f80-c6c3-11eb-8b91-75b5433d8a33.png">

alright now I'm stumped. if I download it through the link here via HTTPS, I get the same error as you. however, if I transfer it from my hosting via FTP using FileZilla to my Desktop folder and unzip it there, it works fine. I also compared md5 hashes between both zip files and they're identical. also, if I transfer via FTP to my Downloads folder, I also get this error. downloading via HTTPS to the Desktop doesn't work as well.

It's got to be some annoying bug with code signing verification, but at the moment I'm at a loss how to fix it.

**Comment by YaxinCheng** [comment]
Hey @shermanikk , sorry for bothering you. As someone already opened a ticket #1378 about incomplete characters, it also happened to me. I went through the build process you provided, and it does not seem to be related to character encodings. So I have tried both Chinese and Japanese characters, and it appears to have a missing byte for every second character. Do you have any insights about which part might got wrong? Thank you

**Comment by shermanikk** [comment]
> 
> 
> Hey @shermanikk , sorry for bothering you. As someone already opened a ticket #1378 about incomplete characters, it also happened to me. I went through the build process you provided, and it does not seem to be related to character encodings. So I have tried both Chinese and Japanese characters, and it appears to have a missing byte for every second character. Do you have any insights about which part might get wrong? Thank you

It's no problem, but I'm definitely out of my depth here. I've never really done any pascal programming, getting fpc and lazarus set up for transgui was my first exposure to it. 

One thing that might be illuminating is to take a look at the actual characters being output, and compare them to the correct values on a binary level.  It might be a little vs big-endian issue of some kind since before no matter the OS transmission or transgui was running on, they were all using x86-64 CPUs. That isn't true anymore and there may be some assumptions in the transgui code that aren't true anymore because of that. If there's an arm64 native version of Transmission I'll try connecting to it and see if the encoding is still screwed up.

Anyways, it's definitely an odd issue and if someone with more experience with pascal and the transgui source wants to help it would be greatly appreciated!

**Comment by tcurdt** [comment]
Thanks for all the great progress here.

While the build from @shermanikk starts up fine, it does still seem to have an openssl issue?

<img width="522" alt="Screenshot 2021-12-03 at 16 22 44" src="https://user-images.githubusercontent.com/13697/144627937-c6754cba-ab2a-453b-9628-2357c6d0c38a.png">

It's also not linked - or is it linked statically?

```
$ otool -L Transmission\ Remote\ GUI.app/Contents/MacOS/transgui 
Transmission Remote GUI.app/Contents/MacOS/transgui:
	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 23.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1770.255.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1770.255.0)
	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
	/System/Library/Frameworks/CoreData.framework/Versions/A/CoreData (compatibility version 1.0.0, current version 1044.3.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2022.20.117)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 54.0.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1463.2.1)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1122.11.0)
	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)
```

**Comment by shermanikk** [comment]
> Thanks for all the great progress here.
> 
> While the build from @shermanikk starts up fine, it does still seem to have an openssl issue?
> 
> <img alt="Screenshot 2021-12-03 at 16 22 44" width="522" src="https://user-images.githubusercontent.com/13697/144627937-c6754cba-ab2a-453b-9628-2357c6d0c38a.png">
> 
> It's also not linked - or is it linked statically?
> 
> ```
> $ otool -L Transmission\ Remote\ GUI.app/Contents/MacOS/transgui 
> Transmission Remote GUI.app/Contents/MacOS/transgui:
> 	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa (compatibility version 1.0.0, current version 23.0.0)
> 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
> 	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
> 	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
> 	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1770.255.0)
> 	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1770.255.0)
> 	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore (compatibility version 1.2.0, current version 1.11.0)
> 	/System/Library/Frameworks/CoreData.framework/Versions/A/CoreData (compatibility version 1.0.0, current version 1044.3.0)
> 	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2022.20.117)
> 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 54.0.0)
> 	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1463.2.1)
> 	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1122.11.0)
> 	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)
> ```

It's definitely compiled, if you look in the MacOS directory you'll see both libcrypto and libssl. Also if you do an otool -l transgui the last line is
>Load command 29
>          cmd LC_RPATH
>      cmdsize 32
>         path @executable_path/. (offset 12)

so it's just looking in it's own directory for the ssl libs. checking in activity monitor when it's open confirms that it is using them as well. I haven't seen that error you're getting before, usually the error I get when I mess up the SSL install is it complaining that it can't find the libs. Do you have homebrew installed with openssl@1.1 built?

**Comment by tcurdt** [comment]
> It's definitely compiled, if you look in the MacOS directory you'll see both libcrypto and libssl. Also if you do an otool -l transgui the last line is
> 
> > Load command 29
> > cmd LC_RPATH
> > cmdsize 32
> > path @executable_path/. (offset 12)
> 
> so it's just looking in it's own directory for the ssl libs. checking in activity monitor when it's open confirms that it is using them as well.

I can confirm they are included:

```
$ find /Applications/Transmission\ Remote\ GUI.app 
/Applications/Transmission Remote GUI.app
/Applications/Transmission Remote GUI.app/Contents
/Applications/Transmission Remote GUI.app/Contents/_CodeSignature
/Applications/Transmission Remote GUI.app/Contents/_CodeSignature/CodeResources
/Applications/Transmission Remote GUI.app/Contents/MacOS
/Applications/Transmission Remote GUI.app/Contents/MacOS/libssl.1.1.dylib
/Applications/Transmission Remote GUI.app/Contents/MacOS/transgui
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.ru
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.ro
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.el
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.lv
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.cs
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.it
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.ca
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.be
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.pt_br
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.no
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.zh_tw
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.da
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.sv
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.pl
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.template
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.tr
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.es
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.fr
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.fi
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.en
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.lt
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.ko
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.de
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.hr
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.nl
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.hu
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.zh
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.uk
/Applications/Transmission Remote GUI.app/Contents/MacOS/lang/transgui.pt
/Applications/Transmission Remote GUI.app/Contents/MacOS/libcrypto.1.1.dylib
/Applications/Transmission Remote GUI.app/Contents/Resources
/Applications/Transmission Remote GUI.app/Contents/Resources/transgui.icns
/Applications/Transmission Remote GUI.app/Contents/Info.plist
/Applications/Transmission Remote GUI.app/Contents/PkgInfo
```
but why aren't they used then?

> I haven't seen that error you're getting before, usually the error I get when I mess up the SSL install is it complaining that it can't find the libs. Do you have homebrew installed with openssl@1.1 built?

I think it was a transitive dep - but it's there

    $ brew list | grep openssl
    openssl@1.1

**Comment by tcurdt** [comment]
How do you see loaded libs in the "Activity Monitor"?

**Comment by shermanikk** [comment]
> How do you see loaded libs in the "Activity Monitor"?

Well not loaded libs necessarily, but if you double click a process it will show you any files the program has open. The dylibs are listed there.

**Comment by tcurdt** [comment]
Ah, ok - I thought I missed something :)

Here are my open files:

```
cwd
/
txt
/Applications/Transmission Remote GUI.app/Contents/MacOS/transgui
txt
/Library/Preferences/Logging/.plist-cache.FDC81N2f
txt
/usr/lib/dyld
txt
/usr/share/icu/icudt68l.dat
txt
/System/Library/Fonts/SFNS.ttf
txt
/System/Library/Fonts/Helvetica.ttc
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/SystemAppearance.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/Aqua.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/FauxVibrantLight.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/VibrantLight.car
txt
/System/Library/Fonts/Menlo.ttc
txt
/private/var/db/analyticsd/events.whitelist
txt
/usr/lib/libobjc-trampolines.dylib
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/FauxVibrantDark.car
txt
/System/Library/CoreServices/CoreGlyphs.bundle/Contents/Resources/Assets.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/DarkAqua.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/VibrantDark.car
txt
/System/Library/CoreServices/SystemAppearance.bundle/Contents/Resources/Assets.car
txt
/System/Library/Keyboard Layouts/AppleKeyboardLayouts.bundle/Contents/Resources/AppleKeyboardLayouts-L.dat
txt
/System/Library/Caches/com.apple.IntlDataCache.le.kbdx
txt
/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/Resources/Extras2.rsrc
txt
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/16777235_530/functions.data
txt
/System/Library/Extensions/AGXMetalG13X.bundle/Contents/Resources/ds.g13s
txt
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/31001/libraries.data
txt
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/0/com.apple.LaunchServices.dv/com.apple.LaunchServices-3027-v2.csstore
0
/dev/null
1
/dev/null
2
/dev/null
3
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_stdlib.metallib
4
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/16777235_530/functions.data
5
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/16777235_530/functions.list
6
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_filters.metallib
7
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_stdlib_stitchable.metallib
8
/System/Library/Frameworks/CoreImage.framework/Versions/A/Resources/ci_filters_stitchable.metallib
9
/Users/tcurdt/Library/Saved Application State/com.transgui.savedState/data.data
10
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/31001/libraries.data
11
/private/var/folders/2j/qlgsy9ys335cdkc03vd8byfr0000gn/C/com.transgui/com.apple.metal/31001/libraries.list
12
/Users/tcurdt/Library/Saved Application State/com.transgui.savedState/windows.plist
13
/Users/tcurdt/Library/Saved Application State/com.transgui.savedState/window_2.data
```

**Comment by shermanikk** [comment]
Have you tried compiling your own build of transgui using the instructions I posted? I know it's a bit of work but that's all I can think of at the moment. Wish I could be of more help but like I've said, I'm hardly a developer and this is my first experience with pascal. Literally everything I know I've posted in this thread.

**Comment by tcurdt** [comment]
> Have you tried compiling your own build of transgui using the instructions I posted? I know it's a bit of work but that's all I can think of at the moment. Wish I could be of more help but like I've said, I'm hardly a developer and this is my first experience with pascal. Literally everything I know I've posted in this thread.

No, I didn't go down that rabbit hole just yet. I have another bot interface that also works OK.
So the priority is not that high for me.

I just got the M1 and still so many projects that need fixing :-(
I was hoping your build could bridge the time to an official release.

I am still a little puzzled on why your build does not work for me. It kind of looks like it should.
At least I am not seeing an obvious problem.

I am wondering if it works for others.

**Comment by frankyw** [comment]
Works fine for me

**Comment by tcurdt** [comment]
> Works fine for me

You were able to update the maxmind db, @frankyw?
I didn't test anything else yet.

**Comment by shermanikk** [comment]
> > Have you tried compiling your own build of transgui using the instructions I posted? I know it's a bit of work but that's all I can think of at the moment. Wish I could be of more help but like I've said, I'm hardly a developer and this is my first experience with pascal. Literally everything I know I've posted in this thread.
> 
> No, I didn't go down that rabbit hole just yet. I have another bot interface that also works OK. So the priority is not that high for me.
> 
> I just got the M1 and still so many projects that need fixing :-( I was hoping your build could bridge the time to an official release.
> 
> I am still a little puzzled on why your build does not work for me. It kind of looks like it should. At least I am not seeing an obvious problem.
> 
> I am wondering if it works for others.

One thing that will make it easier to build it yourself is that you no longer need to compile FPC for aarch64 yourself, you can simply download the prebuilt version from [here](https://phoenixnap.dl.sourceforge.net/project/freepascal/Mac%20OS%20X/3.2.2/fpc-3.2.2.intelarm64-macosx.dmg). I tried building transgui with it and was successful, including loading the openssl libs from homebrew.

You can try that build (using FPC 3.2.2 and Lazarus 2.3.0 trunk) I made here: [transgui-macosx-arm64.v2.zip](https://shermanikk.net/stuff/transgui-macosx-arm64.v2.zip)

**Comment by ijaureguialzo** [comment]
I've just tested `transgui-macosx-arm64.v2.zip`and it works fine for me on M1, even the GeoIP download which the previous one didn't.

**Comment by ohld** [comment]
I found this link to macos builds of Transmission. Downloaded the latest release and it works fine! MacBook Pro 14' M1.

https://build.transmissionbt.com/job/trunk-mac/

**Comment by ijaureguialzo** [comment]
> I found this link to macos builds of Transmission. Downloaded the latest release and it works fine! MacBook Pro 14' M1.
> 
> https://build.transmissionbt.com/job/trunk-mac/

That's another program, not the one we are discussing here...

**Comment by shermanikk** [comment]
If anyone was having issue #1378 I rebuilt transgui with fpc 3.3.1 and lazarus 2.3.0 and kanji/chinese/korean characters are showing up correctly again. Not sure what the deal was but hey, it works! 

Same link as the old build, and I make no promises or guarantees this won't cause bodily injury or constitute a fire hazard: https://shermanikk.net/stuff/transgui-macosx-arm64.v2.zip

**Comment by alfenheim** [comment]
> If anyone was having issue #1378 I rebuilt transgui with fpc 3.3.1 and lazarus 2.3.0 and kanji/chinese/korean characters are showing up correctly again. Not sure what the deal was but hey, it works!
> 
> Same link as the old build, and I make no promises or guarantees this won't cause bodily injury or constitute a fire hazard: https://shermanikk.net/stuff/transgui-macosx-arm64.v2.zip

The link is not working anymore to download, can you please upload/share again?

**Comment by shermanikk** [comment]
> > If anyone was having issue #1378 I rebuilt transgui with fpc 3.3.1 and lazarus 2.3.0 and kanji/chinese/korean characters are showing up correctly again. Not sure what the deal was but hey, it works!
> > Same link as the old build, and I make no promises or guarantees this won't cause bodily injury or constitute a fire hazard: https://shermanikk.net/stuff/transgui-macosx-arm64.v2.zip
> 
> The link is not working anymore to download, can you please upload/share again?

Apologies, edited my post with the correct link.

**Comment by flying-sausages** [comment]
@shermanikk could you try rebuild it with these patches please? #1306

**Comment by aniqueta** [comment]
@shermanikk Trying this a few months later than others here. I have the the above SSL error that @tcurdt has experienced. Environment does not have Homebrew, and reading this thread it seems openssl should be compiled with the build? Would appreciate an Apple Silicon build. Thanks!

**Comment by shermanikk** [comment]
> @shermanikk Trying this a few months later than others here. I have the the above SSL error that @tcurdt has experienced. Environment does not have Homebrew, and reading this thread it seems openssl should be compiled with the build? Would appreciate an Apple Silicon build. Thanks!

I'm glad you said something, I didn't realize it still wasn't working properly! So, it turns out that libssl.1.1.dylib included an LC_LOAD_DYLIB command that was still referencing the original install location of libcrypto in the homebrew directory, and that was why despite the fact that transgui was loading both libcrypto.1.1.dylib and libssl.1.1.dylib from it's own directory, it still required the homebrew install location to have libcrypto.1.1.dylib (/opt/homebrew/Cellar/openssl@1.1/1.1.1o/lib/libcrypto.1.1.dylib in this case). So, after a quick few edits using install_name_tool to both dylibs and a resigning of the entire .app bundle, I believe it should be working without requiring any homebrew installs!

I've updated the .zip linked before, but here it is again: https://shermanikk.net/stuff/transgui-macosx-arm64.zip

Please let me know if this works for you, it's built using fpc 3.3.1 and lazarus 2.3.0 on MacOS 12.3.1 on a 2020 M1 MacBook Air.

**Comment by shermanikk** [comment]
@tcurdt @flying-sausages 

I know it's been a while, but I have finally figured out the OpenSSL issue, see the above post! Try downloading my latest build and see if it's working for you.

**Comment by nickv2002** [comment]
Seems like @shermanikk's new build is working great for me at least.

Can we get a GitHub PR to get this back into the main project or, barring that, put together a fork?

Also, I have an Apple dev account I use for signing open source software with Apple (see [Simple Comic](https://github.com/MaddTheSane/Simple-Comic/releases/tag/App-Store-1.9.2)) so LMK if it would help to have signed binaries so folks can launch them more easily/safely.

**Comment by jvacek** [comment]
Is there any way the changes necessary to build this could be PR'd into master?

**Comment by isevewel** [comment]
I've just tested but it don't work on m1 macbook pro.
![무제](https://user-images.githubusercontent.com/16564757/171387262-7450105d-ab3e-4e82-b806-907ee3b25ee1.png)

**Comment by shermanikk** [comment]
> I've just tested but it don't work on m1 macbook pro.
> 
> ![무제](https://user-images.githubusercontent.com/16564757/171387262-7450105d-ab3e-4e82-b806-907ee3b25ee1.png)

What version of MacOS are you running? Do you have Homebrew installed?

**Comment by acahir** [comment]
> @tcurdt @flying-sausages
> 
> I know it's been a while, but I have finally figured out the OpenSSL issue, see the above post! Try downloading my latest build and see if it's working for you.

I just tried your latest build on a new MBP running macOS 12.4 and it worked without problems, though I'm not connecting through SSL.

**Comment by adissimo25** [comment]
Hi, I installed my older BackUp from my MBP 2010 to my new MBP m1 and have the following error:
<img width="372" alt="Bildschirmfoto 2022-08-13 um 18 36 26" src="https://user-images.githubusercontent.com/111193780/184502975-9ecdeac9-43ff-4116-aec7-fcd60b5cff57.png">

MacOS 12.5 and Homebrew is installed

Thanks for help

**Comment by acahir** [comment]
I was just setting up a new install, and remembered I had found a native build somewhere. 

Rather than have this built out of band, are the maintainers still interested in a universal build? There's no reason why it has to be arm64 or x86-64 builds, the build script should be able to create a universal bundle with the right settings.

**Comment by nickv2002** [comment]
If we can get a universal version of the app, I can also sign it with my developer account so it can launch without warnings. (As I do for [Simple Comic](https://github.com/MaddTheSane/Simple-Comic/releases/tag/App-Store-1.9.5))

**Comment by acahir** [comment]
Example from Apple's [reference](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary):

```
x86_app: main.c
    $(CC) main.c -o x86_app -target x86_64-apple-macos10.12
arm_app: main.c
    $(CC) main.c -o arm_app -target arm64-apple-macos11
universal_app: x86_app arm_app
    lipo -create -output universal_app x86_app arm_app
```

So instead of changing the build lines as suggested in [Step 6](https://github.com/transmission-remote-gui/transgui/issues/1355#issuecomment-762940021) above, both should be included and then the `lipo` tool can create a universal binary. 

Also note in the references that it's better to build all native libraries into the binaries, to make sure they are built for the correct arch. Not sure if that's required or not, seem like it could use the dynamic libs correctly as long as the correct arch was installed (which HB should take care of already).

**Comment by Willian-Zhang** [comment]
> > @shermanikk Trying this a few months later than others here. I have the the above SSL error that @tcurdt has experienced. Environment does not have Homebrew, and reading this thread it seems openssl should be compiled with the build? Would appreciate an Apple Silicon build. Thanks!
> 
> I'm glad you said something, I didn't realize it still wasn't working properly! So, it turns out that libssl.1.1.dylib included an LC_LOAD_DYLIB command that was still referencing the original install location of libcrypto in the homebrew directory, and that was why despite the fact that transgui was loading both libcrypto.1.1.dylib and libssl.1.1.dylib from it's own directory, it still required the homebrew install location to have libcrypto.1.1.dylib (/opt/homebrew/Cellar/openssl@1.1/1.1.1o/lib/libcrypto.1.1.dylib in this case). So, after a quick few edits using install_name_tool to both dylibs and a resigning of the entire .app bundle, I believe it should be working without requiring any homebrew installs!
> 
> I've updated the .zip linked before, but here it is again: https://shermanikk.net/stuff/transgui-macosx-arm64.zip
> 
> Please let me know if this works for you, it's built using fpc 3.3.1 and lazarus 2.3.0 on MacOS 12.3.1 on a 2020 M1 MacBook Air.

I'm running your build on Ventura 13.0 (22A380). Just found a new issue only after upgrade to Ventura:
When opening the app with associated torrent file, the pop window for creation for new torrent task would go behind the main window, instead of going top front.

**Comment by jata1** [comment]
@shermanikk - just wanted to say thanks. This works fine / native on my M2 MacBook Air with latest Ventura.

Will all of the work you and others have done to get this working be merged into the main dev branch for this project?

**Comment by PeterDaveHello** [comment]
@jvacek Sure, that's a good idea, I'm gonna need to take some time to sort it out, make sure the build and release process can be automatic for the next release.

Anyone looking forward to have this update, please feel free to help us all to organize the information and works here, that may speed up the whole process to get it done.

Thank you all for the waiting and contribution.

**Comment by Smultie** [comment]
Thanks a lot for your work, @shermanikk 
Would be great indeed if the changes would be implemented in the original app.

**Comment by mreid-tt** [comment]
> Thanks a lot for your work, @shermanikk Would be great indeed if the changes would be implemented in the original app.

+1

**Comment by urda** [comment]
Alas this took too long and I've migrated my servers to another torrent platform.

**Comment by mreid-tt** [comment]
> If there's anything else you'd like me to do, I'd be happy to! I have been using transgui every day for years now to control my home torrent server and it's been fantastic. Heaps better than some janky, slow WebUI!

Would it be possible to create a PR to enable the Apple Silicon build?

**Comment by PeterDaveHello** [comment]
It'll be great if any contributors would like to send a pull request.

As you can see, many pull requests have been reviewed and merged recently, but my spare time can be very limited, and I'm still working on some Linux issues. So, if someone can help with this issue together, it would be very much appreciated.

**Comment by jvacek** [comment]
> Later today I will start from scratch with a clean install and do a little step-by-step guide (excluding getting the tools themselves installed since those guides already exist) that includes all the little caveats I learned along the way, especially the subtle issue I had getting the OpenSSL libs working properly.

@shermanikk any updates by any chance?

**Comment by ArneAnka** [comment]
I want to build from the latest merge to get labels/groups working. How can i build it?

**Comment by PeterDaveHello** [comment]
@xavery would you like to send a pull request for it? Thanks a lot!

**Comment by chkuendig** [comment]
I think it was @shermanikk who had managed to setup a build - https://github.com/transmission-remote-gui/transgui/issues/1355#issuecomment-1477993771 and https://github.com/transmission-remote-gui/transgui/issues/1355#issuecomment-1125818588. This thread is a bit of a mess #1358

**Comment by xavery** [comment]
For everyone interested, I have just added Mac Silicon builds to my fork. The first release is [here](https://github.com/xavery/transgui/suites/14655163568/artifacts/831550754) - in fact, the binaries are universal so nothing changes for existing macOS users.

In my very limited testing, it seems no less stable than the Intel version, which is probably as good as it's going to get. If I don't get any overwhelmingly negative feedback over the next few weeks, I'll create a new point release and provide universal binaries for macOS from now on.

@PeterDaveHello Feel free to merge the changes from my fork to this repository, but keep in mind that I made some quite significant changes to how the project is built, i.e. Makefiles have been replaced with lazbuild and all CI stuff is handled by GitHub Actions instead of Travis. Most importantly perhaps, this requires an actual native Apple Silicon hardware to be compiled : I have not tried cross-compiling to M1 from an Intel Mac.

**Comment by Smultie** [comment]
The link shows a 404 for me.

**Comment by xavery** [comment]
> The link shows a 404 for me.

As this is a build artifact and not a "release binary", you need to be logged in with your GitHub account to download. Here's the same file as an attachment, hopefully it works better : [TransGUI macOS Universal Binary.zip](https://github.com/transmission-remote-gui/transgui/files/12207731/TransGUI.macOS.Universal.Binary.zip)

**Comment by Smultie** [comment]
> For everyone interested, I have just added Mac Silicon builds to my fork. The first release is [here](https://github.com/xavery/transgui/suites/14655163568/artifacts/831550754) - in fact, the binaries are universal so nothing changes for existing macOS users.
> 
> In my very limited testing, it seems no less stable than the Intel version, which is probably as good as it's going to get. If I don't get any overwhelmingly negative feedback over the next few weeks, I'll create a new point release and provide universal binaries for macOS from now on, or for as long as I have access to an M1 runner. Please create issues in the [fork](https://github.com/xavery/transgui/issues) instead of here in case of problems.
> 
> @PeterDaveHello Feel free to merge the changes from my fork to this repository, but keep in mind that I made some quite significant changes to how the project is built, i.e. Makefiles have been replaced with lazbuild and all CI stuff is handled by GitHub Actions instead of Travis. ~Most importantly perhaps, this requires actual native Apple Silicon hardware to be compiled : I have not tried cross-compiling to M1 from an Intel Mac.~ There is also support for cross-compiling to M1 from an Intel Mac : the executables created this way were perfectly usable on the M1 Mac that I use for testing.

Maybe for next release, it's an idea to have the Command button work for CTRL-commands?
This is the only app where I need to use the CTRL-button for stuff I normally use the Command button (in this case Reannounce (get more peers)

**Comment by xavery** [comment]
@Smultie Please have a look at https://github.com/xavery/transgui/issues/45 and comment there if you can, thanks.

**Comment by Erebus999** [comment]
News
The GUI Close after added a Torrent File.
MACOS Beta 27

**Comment by nickv2002** [comment]
I did an LLM-powered Swift rewrite that is obviously built for Apple Silicon with some small new features (eg multi-host failover). I signed my builds too so they don't require Gatekeeper hoops to run. [nickv2002/Transmission-Remote-MacOS](https://github.com/nickv2002/Transmission-Remote-MacOS)

I have not tried it on macOS 26 or 27 (I'm still on Sequoia) but since it's trying to be a boring native Swift app using standard AppKit it should work just fine on those (baring any OS bugs).

**Comment by jvacek** [comment]
haha @nickv2002 you beat me to it! I went down the SwiftUI route with my RPC client and core logic separated into packages, and instead went full on with being macos26.0+ only https://github.com/jvacek/TransmissionSwift

**Comment by pramod-bls** [comment]
LOL @nickv2002 , @jvacek I wish i knew this before I created my own. Anyways, I have a very different take.  I wrote one where you can connect to multiple server including transmission, deluge or qbittorrent at the same time and manage them simultaneously. the UI is completely flexible and can be rearranged (I have a very wide screen). If any one is interested, I have releases for mac/win/linux [Torrentdeck](https://github.com/pramod-bls/torrentdeck)

Oh for pesky admins... MacOS the app is signed! that was important for me.

**Comment by nickv2002** [comment]
Cheers to you all. No reason they can't all coexist. I write mine 98% for myself (still running macOS Sequoia) though it hews very close to this original project, glad to see folks with other prefs building OSS.
